<a href="https://colab.research.google.com/github/skj092/experiments/blob/main/Activation_function_and_Weight_initializer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Objective : 

**To Study different activation functions and weight initializers**

In [2]:
!pip install mlflow -q

     |████████████████████████████████| 14.2MB 209kB/s 
     |████████████████████████████████| 153kB 44.2MB/s 
     |████████████████████████████████| 174kB 34.5MB/s 
     |████████████████████████████████| 61kB 8.3MB/s 
     |████████████████████████████████| 1.1MB 33.5MB/s 
     |████████████████████████████████| 81kB 9.7MB/s 
     |████████████████████████████████| 71kB 7.6MB/s 
     |████████████████████████████████| 71kB 8.9MB/s 
     |████████████████████████████████| 81kB 9.6MB/s 


In [3]:
# TensorFlow and tf.keras
import tensorflow as tf

# Helper libraries
import numpy as np
from tensorflow.keras import initializers
from tensorflow.python.keras import activations

print(tf.__version__)

# downloading fashion_mnist data
fashion_mnist = tf.keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

train_images = train_images / 255.0

test_images = test_images / 255.0              


import mlflow.tensorflow
mlflow.tensorflow.autolog()

2.5.0


2021/06/13 10:06:24 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of tensorflow. If you encounter errors during autologging, try upgrading / downgrading tensorflow to a supported version, or try upgrading MLflow.


# With Default Weight Initializer

## 1. relu

In [4]:
activation = tf.keras.activations.relu

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)
    mlflow.tensorflow.mlflow.log_param('activation', activation)

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten (Flatten)            (None, 784)               0         
_________________________________________________________________
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 7s 2ms/step - loss: 0.4926 - accuracy: 0.8264
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3727 - accuracy: 0.8650
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3343 - accuracy: 0.8782
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - loss

## 2. Softmax

In [5]:
activation = tf.keras.activations.softmax

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('initializer', model.layers[1].kernel_initializer)
    mlflow.tensorflow.mlflow.log_param('activation', activation)

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_1 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_3 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 1.4005 - accuracy: 0.6022
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.8426 - accuracy: 0.6703
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.7244 - accuracy: 0.6929
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - lo

## 3. Sigmoid

In [6]:
activation = tf.keras.activations.sigmoid

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('initializer', model.layers[1].kernel_initializer)
    mlflow.tensorflow.mlflow.log_param('activation', activation)

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_2 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_5 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.5435 - accuracy: 0.8143
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3905 - accuracy: 0.8595
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3541 - accuracy: 0.8714
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - lo

## 4. tanh

In [7]:
activation = tf.keras.activations.tanh

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('initializer', model.layers[1].kernel_initializer)
    mlflow.tensorflow.mlflow.log_param('activation', activation)

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_3 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_7 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.4751 - accuracy: 0.8273
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3683 - accuracy: 0.8657
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3339 - accuracy: 0.8783
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - lo

## 5. selu

In [8]:
activation = tf.keras.activations.selu

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('initializer', model.layers[1].kernel_initializer)
    mlflow.tensorflow.mlflow.log_param('activation', activation)

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_4 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_9 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.4932 - accuracy: 0.8245
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3913 - accuracy: 0.8572
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3523 - accuracy: 0.8706
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - lo

# With Weight Initializer

## 1. relu + he_normal

In [9]:
activation = tf.keras.activations.relu
initializer = tf.keras.initializers.he_normal

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_5 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_10 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_11 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.5068 - accuracy: 0.8225
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3815 - accuracy: 0.8630
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3430 - accuracy: 0.8752
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - lo

## 2. softmax + glorot_normal

In [10]:
activation = tf.keras.activations.softmax
initializer = tf.keras.initializers.glorot_normal

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_6 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_12 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_13 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 1.3742 - accuracy: 0.6333
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.8280 - accuracy: 0.6804
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.7182 - accuracy: 0.7108
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - lo

## 3. sigmoid + glorot_mormal

In [11]:
initializer = tf.keras.initializers.glorot_normal
activation = tf.keras.activations.sigmoid

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)

Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_7 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_14 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_15 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.5449 - accuracy: 0.8157
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3913 - accuracy: 0.8583
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3551 - accuracy: 0.8710
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - lo

## 4. tanh + glorot_normal

In [12]:
initializer = tf.keras.initializers.glorot_normal
activation = tf.keras.activations.tanh

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)

Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_8 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_16 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_17 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.4754 - accuracy: 0.8293
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3702 - accuracy: 0.8647
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3357 - accuracy: 0.8766
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - lo

## 5. selu + lucan_normal

In [13]:
activation = tf.keras.activations.selu
initializer = tf.keras.initializers.lecun_normal

initializer = tf.keras.initializers.glorot_normal
activation = tf.keras.activations.tanh

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)

Model: "sequential_9"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_9 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_18 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_19 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.4785 - accuracy: 0.8277
Epoch 2/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3716 - accuracy: 0.8648
Epoch 3/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3358 - accuracy: 0.8776
Epoch 4/10
1875/1875 [==============================] - 5s 2ms/step - lo

# Results : 

|Act Func|initializer|train_loss|train_accuracy|test_loss|test_accuracy|
|----|----|----|----|----|----|
|relu|GlorotUniform|0.2378 |0.9109	|0.3364 |0.8807|
|relu|he_normal|0.2423 |0.9089	|0.3358 |0.8867|
|softmax|GlorotUniform|0.4847 |0.8383	|0.5149 |0.8275|
|softmax|glorot_normal|0.4730 |0.8419	|0.5077 |0.8318|
|sigmoid|GlorotUniform|0.2503 |0.9078	|0.3451 |0.8767|
|sigmoid|glorot_normal|0.2517 |0.9063	|0.3325 |0.8791|
|tanh|GlorotUniform|0.2396 |0.9101	|0.3354 |0.8812|
|tanh|glorot_normal|0.2433 |0.9086	|0.3297 |0.8811|
|selu|GlorotUniform|0.2470 |0.9064	|0.3470 |0.8791|
|selu|lucan_normal|0.2416 |0.9112	|0.3346 |0.8812|

**Observations**
- By Default Dense layer have `Glorot Uniform` weight initializer.
- `Softmax` performed `worst` with both the initializers.
- Default weight initializer (GlorotUniform) performed worse in every activation function.
- Order of model according to test_loss is : `tanh > selu > sigmoid > relu > softmax`.
- Order of model according to test_accuracy is : `relu > tanh > sigmoid > sigmoid > softmax`.